plumbing/object/commitnode.go

```diff
@@ -0,0 +1,94 @@
+package object
+
+import (
+	"io"
+	"time"
+
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/plumbing/storer"
+)
+
+// CommitNode is generic interface encapsulating a lightweight commit object retrieved
+// from CommitNodeIndex
+type CommitNode interface {

```

Comment:
Is it possible to get the generation of the commit? This information is interesting to compare two commits or to stop walking at a certain level.

Reply:
It is available at the format level, but I didn't propagate it here and it is not exposed through API at this level. I will look into it. One thing to consider though is that the Git folks are looking into replacing generation numbers with "corrected commit date". I will link to the relevent discussion once I get back to my computer...

Reply:
This? https://public-inbox.org/git/6367e30a-1b3a-4fe9-611b-d931f51effef@gmail.com/

Reply:
Yep, V3 was the one I most recently read about. There's no final decision about any particular approach, but I didn't feel confident about designing API around something that is likely to change.

Reply:
Looks like there are three things we can expose with a reasonable level of forward compatibility with the new proposals:

- `Generation() uint64` API on `CommitNode` that will uphold the property that if `A.Generation() < B.Generation()` then  `B` is unreachable by `A`. While the current format allows the generation to be only 30-bit integer extending it to 64-bit should account for any of the other proposals to work with the same API (except maybe the FELINE version), including the 34-bit corrected commit dates.
- A high-level API like `IsUnreachableFrom(a, b CommitNode)`. The idea is to abstract away the concept of the generations from the user and just expose the most useful underlying property. This may need some thinking since we basically have three states: a) definitely unreachable (property coming from the generations) b) definitely reachable (not in the commit-graph, but perhaps something that is stored in reachability bitmap for some subset of commits?) c) we don't know.
- High-level API for priority heaps for traversing the graph in topological order by reusing the Generation property without directly exposing it.

Thoughts? I don't feel like implementing the last one since I currently have no use case for it, but maybe it would be useful for merge-base or something of that sort.

Reply:
I didn't feel confident enough to establish API shape for two of the above options. I've added the `Generation()` method to `CommitNode` in a fashion similar to what `git` uses internally.